### PR TITLE
Finished initial refactor.

### PR DESCRIPTION
**Pull request by tritoke**
Also fixed a bug where $? wasn't expanded properly.
And generally attempted to make the code more idiomatic.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -4,7 +4,10 @@
 
 # Remove Cargo.lock from gitignore if creating an executable, leave it for libraries
 # More information here https://doc.rust-lang.org/cargo/guide/cargo-toml-vs-cargo-lock.html
-# Cargo.lock
+Cargo.lock
```
Comment (chorman0773) [inline .gitignore]: Cargo.lock should be checked out for executables.

**Review by chorman0773 — CHANGES_REQUESTED** [review]
Cargo.lock should be checked out for executables. Otherwise, looks good

File: .gitignore (hunk 1 above)
Comment (tritoke) [reply .gitignore]: sorry I'm not sure I understand, should it be included or not?

Comment (chorman0773) [reply .gitignore]: It should not be in `.gitignore`.

Comment (chorman0773) [reply .gitignore]:
```suggestion
```

Comment (tritoke) [reply .gitignore]: Fixed in commit #3cdd8e5

**Review by chorman0773 — APPROVED** [review]

**Pull request merged**